quiz/quiz.go

```diff
@@ -0,0 +1,182 @@
+// This package implements the quiz, with a variation on part 2.  
+// Instead of having an timer for the entire test, it handles a timer 
+// for each question.  This was more challenging, since characters 
+// entered for previous questions had to be properly discarded
+// to get the desired behavior.
+
+package main
+
+import (
+	"bufio"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"time"
+)
+
+type QuizItem struct {
+	Question string
+	Answer   string
+}
+
+type CharWithTime struct {
+	char string
+	time time.Time
+}
+
+type ErrTimeout struct {
+	message string
+}
+
+func (e *ErrTimeout) Error() string {
+	return e.message
+}
+
+func NewErrorTimeout(message string) *ErrTimeout {
+	return &ErrTimeout{message: message}
+}
+
+func OldReadQuizFromFile(filepath string) []QuizItem {
+	file, err := os.Open(filepath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	items := make([]QuizItem, 0, 100)
+	fscanner := bufio.NewScanner(file)
+	for fscanner.Scan() {
+		line := fscanner.Text()
+		// fmt.Println(line)
+		tokens := strings.Split(line, ",")
+		items = append(items, QuizItem{Question: tokens[0], Answer: tokens[1]})
+	}
+	return items
+}
+
+func ReadQuizFromFile(filepath string) []QuizItem {
+	file, err := os.Open(filepath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	items := make([]QuizItem, 0, 100)
+	r := csv.NewReader(file)
+	for {
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		fmt.Println(record)
+		items = append(items, QuizItem{Question: record[0], Answer: record[1]})
+	}
+	return items
+}
+
+func ReadCharsWithTime(chars chan CharWithTime) {
+	reader := bufio.NewReader(os.Stdin)
+	for {
+		r, _, err := reader.ReadRune()
+		if err != nil {
+			log.Fatal(err)
+		}
+		chars <- CharWithTime{char: string(r), time: time.Now()}
+	}
+}
+
+func ReadStringWithTimeout(chars chan CharWithTime, timeout time.Duration) (string, error) {
+	start := time.Now()
+	// end := start.Add(timeout)
+	result := ""
+	for {
+		select {
+		case c := <-chars:
+			if c.time.Before(start) {
+				// Just ignore old input
+				continue
+			}
+			if c.char == "\n" {
+				// Found end of string
+				return result, nil
+			}
+			// Otherwise add character to result buffer.
+			result += c.char
+		case <-time.After(timeout):
```

Comment:
As nice as this looks, in most of are use cases you will want to use the `time.Timer` so you can defer the close method.

Reply:
I'm not sure what close I would be deferring.  This is reading the channel, so it is not responsible for closing the channel.